src/permission/DefaultPermissions.php

```diff
@@ -24,95 +24,172 @@
+ * @method static Permission COMMAND_TITLE_SELF()
+ * @method static Permission COMMAND_TRANSFERSERVER()
+ * @method static Permission COMMAND_UNBAN_IP()
+ * @method static Permission COMMAND_UNBAN_PLAYER()
+ * @method static Permission COMMAND_VERSION()
+ * @method static Permission COMMAND_WHITELIST_ADD()
+ * @method static Permission COMMAND_WHITELIST_DISABLE()
+ * @method static Permission COMMAND_WHITELIST_ENABLE()
+ * @method static Permission COMMAND_WHITELIST_LIST()
+ * @method static Permission COMMAND_WHITELIST_RELOAD()
+ * @method static Permission COMMAND_WHITELIST_REMOVE()
+ * @method static Permission COMMAND_XP_OTHER()
+ * @method static Permission COMMAND_XP_SELF()
+ * @method static Permission GROUP_CONSOLE()
+ * @method static Permission GROUP_OPERATOR()
+ * @method static Permission GROUP_USER()
+ */
+final class DefaultPermissions{
+	use RegistryTrait;
+
+	public const ROOT_CONSOLE = "pocketmine.group.console";
+	public const ROOT_OPERATOR = "pocketmine.group.operator";
+	public const ROOT_USER = "pocketmine.group.user";
 
 	/**
 	 * @param Permission[] $grantedBy
 	 * @param Permission[] $deniedBy
 	 */
-	public static function registerPermission(Permission $candidate, array $grantedBy = [], array $deniedBy = []) : Permission{
+	private static function registerPermission(Permission $candidate, array $grantedBy = [], array $deniedBy = []) : Permission{
 		foreach($grantedBy as $permission){
-			$permission->addChild($candidate->getName(), true);
+			$permission->addChild($candidate, true);
 		}
 		foreach($deniedBy as $permission){
-			$permission->addChild($candidate->getName(), false);
+			$permission->addChild($candidate, false);
 		}
 		PermissionManager::getInstance()->addPermission($candidate);
 
+		$name = str_replace("pocketmine.", "", $candidate->getName());
+		self::_registryRegister(str_replace(".", "_", $name), $candidate);
+
 		return PermissionManager::getInstance()->getPermission($candidate->getName());
 	}
 
+	protected static function setup() : void{
+		self::registerCorePermissions();
+	}
+
 	public static function registerCorePermissions() : void{
-		$consoleRoot = self::registerPermission(new Permission(self::ROOT_CONSOLE, l10n::pocketmine_permission_group_console()));
-		$operatorRoot = self::registerPermission(new Permission(self::ROOT_OPERATOR, l10n::pocketmine_permission_group_operator()), [$consoleRoot]);
-		$everyoneRoot = self::registerPermission(new Permission(self::ROOT_USER, l10n::pocketmine_permission_group_user()), [$operatorRoot]);
+		$consoleRoot = self::registerPermission(new Permission("pocketmine.group.console", l10n::pocketmine_permission_group_console()));
+		$operatorRoot = self::registerPermission(new Permission("pocketmine.group.operator", l10n::pocketmine_permission_group_operator()), [$consoleRoot]);
+		$everyoneRoot = self::registerPermission(new Permission("pocketmine.group.user", l10n::pocketmine_permission_group_user()), [$operatorRoot]);
```

Comment:
why removing the use of the const ?

Reply:
Constants aren't needed because the permissions will be referenced by registry. It's no different to the likes of VanillaItems and VanillaBlocks.

Reply:
The ROOT_* constants at the top of this class can then be deleted?

Reply:
Probably yes, although I didn't look closely if they were still used anywhere

Reply:
In any case, if I've understood the PR correctly, if you have to use const, you're doing it wrong because the aim of this PR is to remove the use of string in favour of the Permission object.
As far as I'm concerned, const should be removed, to be confirmed.

Reply:
Yep